### PR TITLE
Feature/add python object createion for unboxing sycl event and queue

### PR DESCRIPTION
**Pull request by ZzEeKkAa**
Add python object creation if `dpctl.SyclEvent` or `dpctl.SyclQueue` where allocated in dpjit function.

Closes: #1181 
Closes: #1182 
Closes: #1183

- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

File: numba_dpex/dpnp_iface/_intrinsic.py [hunk 1]
```diff
@@ -1077,3 +1078,59 @@ def codegen(context, builder, sig, args):
...
+        # so it does not steel reference.
+        #
+        # We can avoid it by:
+        #  queue_ref_copy = sycl.dpctl_queue_copy(builder, queue_ref) #noqa E800
+        #  queue_struct_proxy.queue_ref = queue_ref_copy #noqa E800
+        #  queue_struct->meminfo =
+        #     nrt->manage_memory(queue_ref_copy, DPCTLEvent_Delete);
+        # but it will allocate new meminfo object which can negatively affect
+        # performance.
+        # Speaking philosifically attribute is a part of the object and as long
```
Comment (diptorupd) [inline numba_dpex/dpnp_iface/_intrinsic.py:1123]: `s/philosifically/philosophically/g`

File: numba_dpex/dpnp_iface/_intrinsic.py [hunk 2]
```diff
@@ -1077,3 +1078,59 @@ def codegen(context, builder, sig, args):
...
+        # as nobody can still the reference it is a part of the owner object
+        # and lifetime is tighted to it.
```
Comment (diptorupd) [inline numba_dpex/dpnp_iface/_intrinsic.py:1125]: `s/tighted/tied/g`

File: numba_dpex/dpnp_iface/_intrinsic.py [hunk 3]
```diff
@@ -1077,3 +1078,59 @@ def codegen(context, builder, sig, args):
...
+
+        queue_value = queue_struct_proxy._getvalue()
+
+        # We need to incref meminfo so that queue model is preventing parent
+        # ndarray from being destroyed, that can destroy queue that we are
```
Comment (diptorupd) [inline numba_dpex/dpnp_iface/_intrinsic.py:1130]: s/ndarray/SyclQueue/g` ?

Comment (diptorupd) [reply numba_dpex/dpnp_iface/_intrinsic.py:1130]: Oh wait this is an attribute of ndarray. Never mind. Sorry.

File: numba_dpex/dpnp_iface/_intrinsic.py [hunk 4]
```diff
@@ -1077,3 +1078,59 @@ def codegen(context, builder, sig, args):
...
 
     return signature, codegen
+
+
+@intrinsic
+def ol_dpnp_nd_array_sycl_queue(
+    ty_context,
+    ty_dpnp_nd_array: DpnpNdArray,
+):
+    if not isinstance(ty_dpnp_nd_array, DpnpNdArray):
```
Comment (diptorupd) [inline numba_dpex/dpnp_iface/_intrinsic.py:1088]: When can this happen? Should a TypingError be raised here?

Comment (ZzEeKkAa) [reply numba_dpex/dpnp_iface/_intrinsic.py:1087]: It may happen if user calls it directly. It can not happen if called from overload

Comment (ZzEeKkAa) [reply numba_dpex/dpnp_iface/_intrinsic.py:1087]: Added error TypeError raise

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py [hunk 5]
```diff
@@ -139,3 +139,22 @@ def func(shape, queue):
...
+    @dpjit
+    def func(a):
+        queue = a.sycl_queue
+        return dpnp.empty(10, sycl_queue=queue)
+
+    a = dpnp.empty(10)
+    b = func(a)
+
+    assert b is not None
+    assert b.sycl_queue is not None
```
Comment (diptorupd) [inline numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py:160]: we should also be able to assert `b.sycl_queue == a.sycl_queue`?

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py [hunk 6]
```diff
@@ -139,3 +139,22 @@ def func(shape, queue):
     with pytest.raises(errors.TypingError):
         queue = dpctl.SyclQueue()
         func(10, queue)
+
+
+def test_dpnp_empty_with_dpjit_queue():
+    """Tests basic boxing and unboxing of a dpnp.ndarray object.
```
Comment (diptorupd) [inline numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py:145]: The docstring comment does not seem to be accurate.

File: numba_dpex/tests/core/types/DpctlSyclQueue/test_box.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import dpnp
+import pytest
+from dpctl import SyclQueue
+
+from numba_dpex import dpjit
+
+
+@pytest.mark.xfail(reason="dpjit allocates new dpctl.SyclQueue on boxing")
+def test_boxing_without_parent():
+    """Tests basic boxing and unboxing of a dpnp.ndarray object.
```
Comment (diptorupd) [inline numba_dpex/tests/core/types/DpctlSyclQueue/test_box.py:18]: The docstring does not seem to match the test

File: numba_dpex/tests/core/types/DpctlSyclQueue/test_box.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        queue = arr.sycl_queue
+        return queue
+
+    arr = dpnp.empty(10)
+
+    q = func(arr)
+
+    assert q is not None
+    assert isinstance(q, SyclQueue)
+    assert id(q) == id(arr.sycl_queue)  # this must work after dpjit fix
```
Comment (diptorupd) [inline numba_dpex/tests/core/types/DpctlSyclQueue/test_box.py:35]: Only the last assert is xfail right? The other asserts should pass?

File: numba_dpex/dpnp_iface/arrayobj.py [hunk 9]
```diff
@@ -1085,3 +1086,11 @@ def getitem_arraynd_intp(context, builder, sig, args):
         ret = builder.insert_value(ret, sycl_queue_attr, sycl_queue_attr_pos)
 
     return ret
+
+
+@overload_attribute(DpnpNdArray, "sycl_queue")
+def dpnp_nd_array_sycl_queue(arr):
```
Comment (diptorupd) [inline numba_dpex/dpnp_iface/arrayobj.py:1092]: Doc string?

File: numba_dpex/dpnp_iface/_intrinsic.py [hunk 10]
```diff
@@ -1077,3 +1077,61 @@ def codegen(context, builder, sig, args):
...
+        raise TypeError("Argument must be DpnpNdArray")
```
Comment (diptorupd) [inline numba_dpex/dpnp_iface/_intrinsic.py:1088]: I meant `numba.core.errors.TypingError`

File: numba_dpex/tests/core/types/DpctlSyclQueue/test_box.py (hunk 8 above)
Comment (ZzEeKkAa) [reply numba_dpex/tests/core/types/DpctlSyclQueue/test_box.py:35]: Yes, is there a way to achieve it?

Comment (diptorupd) [reply numba_dpex/tests/core/types/DpctlSyclQueue/test_box.py:35]: You will have to write two separate test cases.

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py (hunk 5 above)
Comment (ZzEeKkAa) [reply numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py:160]: It will fail then. I don't know how to add just one assert as expected failure

Comment (diptorupd) [reply numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py:160]: Why will it fail? The `__eq__` of the `dpctl.SyclQueue` class just calls the `DPCTLQueue_Equal` function that should evaluate to true if the `qref` pointers point to the same underlying `sycl::queue` object.

File: numba_dpex/tests/core/types/DpctlSyclQueue/test_box.py (hunk 8 above)
Comment (ZzEeKkAa) [reply numba_dpex/tests/core/types/DpctlSyclQueue/test_box.py:35]: I've been thinking about what test should actually check and updated tests a little bit. So this one is always pass, and another one is failing. Thx

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py (hunk 5 above)
Comment (ZzEeKkAa) [reply numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py]: We must be more specific. It suppose to be same object. Marked test as XFAIL.

File: numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py [hunk 11]
```diff
@@ -139,3 +139,33 @@ def func(shape, queue):
...
+
+    @dpjit
+    def func(a):
+        queue = a.sycl_queue
+        return dpnp.empty(10, sycl_queue=queue)
+
+    a = dpnp.empty(10)
+    b = func(a)
+
+    assert a.sycl_queue == b.sycl_queue
```
Comment (diptorupd) [inline numba_dpex/tests/dpjit_tests/dpnp/test_dpnp_empty.py:171]: 👍

**Review by diptorupd — APPROVED** [review]
Thank you!

**Pull request merged**